src/MessageHandler/SearchNoHitsMessageHandler.php

```diff
@@ -45,151 +45,312 @@ public function __construct(
+    {
+        $message = new VendorImageMessage();
+        $message->setOperation($operation)
+            ->setIdentifier($source->getMatchId())
+            ->setVendorId($source->getVendor()->getId())
+            ->setIdentifierType($source->getMatchType());
+        $this->bus->dispatch($message);
+    }
+
+    /**
+     * Persist a matching "Source" from an UnverifiedVendorImageItem.
+     *
+     * @param UnverifiedVendorImageItem $item
+     *
+     * @return Source
+     *
+     * @throws DBALException
+     */
+    private function persistSource(UnverifiedVendorImageItem $item): Source
+    {
+        if (!$item->isFound()) {
+            throw new \InvalidArgumentException('A "Source" should not be persisted from UnverifiedVendorImageItem that is not "found"');
+        }
+
+        $sourceRepository = $this->em->getRepository(Source::class);
+
+        // There may exist a race condition when multiple queues are
+        // running. To ensure we don't insert duplicates we need to
+        // wrap our search/update/insert in a transaction.
+        $this->em->getConnection()->beginTransaction();
+
+        try {
+            /** @var Source $source */
+            $source = $sourceRepository->findOneBy([
+                'matchId' => $item->getIdentifier(),
+                'matchType' => $item->getIdentifierType(),
+                'vendor' => $item->getVendor(),
+            ]);
+
+            if (is_null($source)) {
+                $source = new Source();
+                $source->setVendor($item->getVendor());
+                $source->setMatchId($item->getIdentifier());
+                $source->setMatchType($item->getIdentifierType());
+                $source->setDate(new \DateTime());
+
+                $this->em->persist($source);
+            }
+
+            $source->setOriginalFile($item->getOriginalFile());
+            $source->setOriginalContentLength($item->getOriginalContentLength());
+            $source->setOriginalLastModified($item->getOriginalLastModified());
+
+            $this->em->flush();
+            $this->em->getConnection()->commit();
+
+            return $source;
+        } catch (DBALException $dbalException) {
+            $this->em->getConnection()->rollBack();
+
+            throw $dbalException;
+        }
+    }
+
+    /**
+     * Try to search the datawell and match source entity.
+     *
+     * @param Material $material
+     *
+     * @return bool
+     */
+    private function mapDatawellSearch(Material $material): bool
+    {
+        $found = false;
+
+        $sourceRepository = $this->em->getRepository(Source::class);
+
+        foreach ($material->getIdentifiers() as $is) {
+            $source = $sourceRepository->findOneByVendorRank($is->getType(), $is->getId());
+
+            // If we have a 'source' that match the material from the datawell we create the relevant jobs
+            // to re-index the source entities
```

Comment:
Missing "." at the end of the comment :-)

Reply:
Fixed